SM dev - SM datamodel and CRUD skeleton

bowenlan-amzn opened the pull request:
Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

*Issue #, if available:*
#280 

*Description of changes:*

## Data model

<details>
<summary>snapshot management policy</summary>

```json
{
    "name": "daily-snapshot",
    "id": "daily-snapshot-sm",
    "description": "Daily snapshot",
    "creation": {
        "schedule": {
            "cron": {
                "expression": "* * * * *",
                "timezone": "America/Los_Angeles"
            }
        },
        "timeout": "1h" // if snapshot creation time is longer than this, skip to next schedule
    },
    "deletion": {
        "schedule": {
            "cron": {
                "expression": "*/2 * * * *",
                "timezone": "America/Los_Angeles"
            }
        },
        "timeout": "1h", // if snapshot deletion time is longer than this, skip to next schedule
        "condition": {
            "max_count": 50, // max number of snapshots kept under this policy
            "max_age": "7d", // max age of snapshots kept under this policy
            "min_count": 5 // min number of snapshots kept under this policy, regardless of max_age condition
        }
    },
    "snapshot_config": {
        "date_expression": "{now/d}", // customize the date format of date shown in snapshot name
        "indices": "*", // other fields in snapshot_config are params for native create snapshot API
        "repository": "repo",
        "ignore_unavailable": "",
        "include_global_state": "",
        "partial": "",
        "metadata": {}
    },
    "last_updated_time": 1648694952,
    "enabled": true, // whether the policy is running or not
    "enabled_time": 1648694952,
    "schedule": { // the schedule to run this policy
      "interval": {
        "period": 1,
        "unit": "Minutes",
        "start_time": 1602100553
      }
    },
    "seq_no": 2,
    "primary_term": 1
}
```
</details>

<details>
<summary>snapshot management metadata</summary>

```json
{
    "policy_seq_no": 2, // policy seqNo and primaryTerm used to find out if policy has been changed
    "policy_primary_term": 1,
    "current_state": "create_condition_met", // the current state of state machine
    "info": { // user facing information
        "pre_condition_failed": "cluster is red, will not create snapshot now.",
        "exception": "[2022-03-25]: Caught exception while...",
        "last_success": "",
        "last_failure": ""
    },
    "creation": {
        "trigger": { // recurring actions trigger
            "next_exeution_time": 1602100553
        },
        "started": { // contains the newest snapshot info that has been started to create
          "name": "daily-snapshot-002",
          "start_time": 1602100553 // used to decide whether timeouts
        },
        "finished": { // record the info of last created snapshot
          "name": "daily-snapshot-001",
          "start_time": 1602100553,
          "end_time": 1602100553
        }
    },
    "deletion": {
        "trigger": {
            "next_exeution_time": 1602100553
        },
        "started": [ // list of snapshots info that has been started to delete
          {
            "name": "daily-snapshot-001",
            "start_time": 1602100553,
            "end_time": 1602100553
          }
        ], 
        "started_time": 1602100553 // the time when last deletion started, used to decide whether timeouts
    },
    "id": "daily-snapshot-sm-metadata",
    "seq_no": 2,
    "primary_term": 1
}
```
</details>

## APIs

Pour in draft for several APIs, not the finalized version, just the basic structure.

- Create policy
- Update policy
- Get policy
- Delete policy

*CheckList:*
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/index-management/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

bowenlan-amzn committed:
Data model and Basic CRUD APIs

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn committed:
Update cached ism config mapping

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn committed:
Re-enable ktlint

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

downsrob commented on src/main/kotlin/org/opensearch/indexmanagement/opensearchapi/OpenSearchExtensions.kt:
```diff
@@ -306,3 +309,35 @@ suspend fun <T> withClosableContext(
+
+inline fun <T> XContentParser.parseArray(block: XContentParser.() -> T): List<T> {
+    val resArr = mutableListOf<T>()
+    if (currentToken() == Token.VALUE_NULL) return resArr
```
This would return an empty array instead of a null value, might be more correct to return null here when there was no array start

bowenlan-amzn committed:
Strengthen scroll search in Coordinator (#356)

* Strengthen scroll search in Coordinator

Normally user won't have >10k ISM jobs, so we don't want to use scroll
search every time.

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

downsrob committed:
Removes rc1 qualifier (#353)

* Removes rc1 qualifier

Signed-off-by: Clay Downs <downsrob@amazon.com>

* Allows error in shrink test

Signed-off-by: Clay Downs <downsrob@amazon.com>

downsrob committed:
Adds workflow to create documentation issues (#358)

* Adds documentation issue workflow

Signed-off-by: Clay Downs <downsrob@amazon.com>

* Makes the issue template more relevant

Signed-off-by: Clay Downs <downsrob@amazon.com>

bowenlan-amzn commented on src/main/kotlin/org/opensearch/indexmanagement/IndexManagementPlugin.kt:
```diff
@@ -208,6 +219,7 @@ class IndexManagementPlugin : JobSchedulerExtension, NetworkPlugin, ActionPlugin
     @Suppress("ComplexMethod")
     override fun getJobParser(): ScheduledJobParser {
         return ScheduledJobParser { xcp, id, jobDocVersion ->
+            log.info("Plugin job parser")
```
How about we use `log.info("sm dev: .....")` for all the logs that probably need to be removed.
And for all the TODO we add in the project, we use `// TODO SM`

bowenlan-amzn replied:
How about we use `log.info("sm dev: .....")` for all the logs that probably need to be removed.
And for all the TODO we add in the project, we use `// TODO SM`

bowenlan-amzn committed:
change log, TODO to SM specific

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

dbbaughe commented on detekt.yml:
```diff
@@ -1,6 +1,6 @@
 # TODO: Remove this before initial release, only for developmental purposes
 build:
-  maxIssues: 20
+  maxIssues: 100
```
🙄

bowenlan-amzn replied:
Still too low? 🤡

Create an issue for this: https://github.com/opensearch-project/index-management/issues/363

dbbaughe commented on src/main/kotlin/org/opensearch/indexmanagement/snapshotmanagement/api/transport/delete/DeleteSMRequest.kt:
```diff
@@ -0,0 +1,27 @@
+import org.opensearch.common.io.stream.StreamInput
+import org.opensearch.common.io.stream.StreamOutput
+
+class DeleteSMRequest(
```
DeleteSMRequest -> DeleteSMPolicyRequest?
Otherwise it's confusing, same w/ all the other classes and variables

dbbaughe commented on src/main/kotlin/org/opensearch/indexmanagement/snapshotmanagement/api/transport/SMActions.kt:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * [TransportIndexSMAction]
+     */
+    const val INDEX_SM_ACTION_NAME = "cluster:admin/opensearch/snapshot_management/write"
```
Make sure this string doesn't conflict w/ any other teams working on snapshot related features.

dbbaughe replied:
*/snapshot_management/*

bowenlan-amzn replied:
I think I get what you are worried about. I want to change it to `cluster:admin/opensearch/snapshot_management/policy/write` to reduce the possible collision with other features.

dbbaughe commented on src/main/resources/mappings/opendistro-ism-config.json:
```diff
@@ -1305,6 +1305,253 @@
         }
       }
+    },
+    "sm": {
```
Do you have a public doc of the data models and usages somewhere?
Hard to grok what all these sub data models are for and to understand them without seeing how they'll be used in implementation.

bowenlan-amzn replied:
Update in this PR description inline, please check

bowenlan-amzn committed:
show build folder, add sourceJar for spi to show source files

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn commented on src/main/kotlin/org/opensearch/indexmanagement/snapshotmanagement/model/SMMetadata.kt:
```diff
@@ -0,0 +1,368 @@
+
+typealias InfoType = Map<String, Any>?
+
+data class SMMetadata(
```
One is for itself. One is for SMPolicy, save the version of SMPolicy in Metadata so that when user change the policy, in the scheduled job we know it got changed and we can address the change.

bowenlan-amzn commented on src/main/kotlin/org/opensearch/indexmanagement/snapshotmanagement/model/SMPolicy.kt:
```diff
@@ -0,0 +1,373 @@
+        init {
+            require(
+                (maxAge != null && minCount != null) || (maxAge == null && minCount == null)
+            ) { "max_age and min_count should exist at the same time." }
```
We want to require user to give a max_count of snapshots to keep.
Max_age is optional, but when user provided, there could be a situation all snapshots are out-dated and should be deleted, but we would want to at least keep some snapshots even they are out-dated.

So max_age and min_count are coupled by design.

However, from user experience perspective, it's not the best to restrict user here but we can deal with it in the inner logic, like a default min_count can be `min of (5, max_count)`.

bowenlan-amzn commented on src/main/kotlin/org/opensearch/indexmanagement/snapshotmanagement/model/SMPolicy.kt:
```diff
@@ -0,0 +1,373 @@
+
+        override fun toXContent(builder: XContentBuilder, params: ToXContent.Params): XContentBuilder {
+            return builder.startObject()
+                .field(MAX_COUNT_FIELD, maxCount)
```
This is by design, we want user to provided a max count so not too many snapshots are created. 

From user experience perspective, we can default this to some value, like 50, not required to provide.

bowenlan-amzn committed:
rename SM to SMPolicy

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn committed:
Unify NO_ID usage

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn committed:
snapshot config

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn committed:
delete condition count age validation

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn committed:
fix NO_ID import in test files

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn committed:
Refactor out the policyName in SMPolicy object

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn committed:
Remove atomic field from SMMetadata, try to do create snapshot transaction without use of it.

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn committed:
ktlint

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn commented:
This commit https://github.com/opensearch-project/index-management/pull/345 on main branch are left out, because DCO problem. But when we merge sm-dev to main, this should be covered.

bowenlan-amzn committed:
update cached ism config mapping

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>

bowenlan-amzn committed:
update config schema version in test

Signed-off-by: bowenlan-amzn <bowenlan23@gmail.com>
